### PR TITLE
Add a bounded lru cache for files

### PR DESCRIPTION
**Pull request by jakemac53**
For now I have set the following defaults, which we can make configurable in the future:

- ~512MB per cache (one for bytes and one for strings)
- 1MB max per file, anything larger than that is never cached

I also fixed an issue where we actually were using the Encoding as a key before but it doesn't implement `hashCode` or `==` so it could have exploded in size since we had separate caches per encoding _instance_. It now uses the name of the Encoding, which should be pretty reasonable.

Note that sizes are based on the file size on disk and not necessarily how much memory is actually consumed by the vm representation of the objects.

Closes https://github.com/dart-lang/build/issues/1461

### CODE REVIEW

File: build_runner_core/lib/src/asset/cache.dart [hunk 1]
```diff
@@ -44,23 +66,76 @@ class CachingAssetReader implements AssetReader {
...
-  Future<List<int>> readAsBytes(AssetId id) =>
-      _bytesContentCache.putIfAbsent(id, () => _delegate.readAsBytes(id));
+  Future<List<int>> readAsBytes(AssetId id, {bool cache = true}) {
+    var cached = _bytesContentCache[id];
+    if (cached != null) return new Future.value(cached);
+
+    return _pendingBytesContentCache.putIfAbsent(
+        id,
+        () => _delegate.readAsBytes(id).then((result) {
+              // Defensive tactic for unawaited futures which could potentially
```
[the remark this answers is not in the document]
Comment (jakemac53) [reply build_runner_core/lib/src/asset/cache.dart]: Basically, if we have unawaited async reads that are still going on after a build completes, this makes sure we don't then insert those potentially stale values into the cache.

However, in this scenario all kinds of other bad things would happen anyways, so I removed the hack here and filed https://github.com/dart-lang/build/issues/1661 instead.

File: build_runner_core/lib/src/asset/cache.dart [hunk 2]
```diff
@@ -17,15 +20,34 @@ import 'package:glob/glob.dart';
...
   /// Cached results of [canRead].
+  ///
+  /// Don't bother using an LRU cache for this since it's just booleans.
   final _canReadCache = <AssetId, Future<bool>>{};
 
   /// Cached results of [readAsString], per [Encoding] type used.
   ///
   /// These are computed and stored lazily using [readAsBytes].
-  final _stringContentCache = <AssetId, Map<Encoding, Future<String>>>{};
+  final _stringContentCache = new LruCache<_AssetIdWithEncoding, String>(
```
[the remark this answers is not in the document]
Comment (jakemac53) [reply build_runner_core/lib/src/asset/cache.dart]: done

File: build_runner_core/lib/src/asset/cache.dart [hunk 3]
```diff
@@ -17,15 +20,34 @@ import 'package:glob/glob.dart';
 /// Does not implement [findAssets].
 class CachingAssetReader implements AssetReader {
   /// Cached results of [readAsBytes].
-  final _bytesContentCache = <AssetId, Future<List<int>>>{};
+  final _bytesContentCache =
+      new LruCache<AssetId, List<int>>(1024 * 1024, 1024 * 1024 * 512, (value) {
+    if (value is Uint8List) {
+      return value.lengthInBytes;
```
[the remark this answers is not in the document]
Comment (jakemac53) [reply build_runner_core/lib/src/asset/cache.dart]: done

File: build_runner_core/lib/src/asset/cache.dart [hunk 4]
```diff
@@ -17,15 +20,29 @@ import 'package:glob/glob.dart';
...
   ///
   /// These are computed and stored lazily using [readAsBytes].
-  final _stringContentCache = <AssetId, Map<Encoding, Future<String>>>{};
+  ///
+  /// Only files read with [utf8] encoding (the default) will ever be cached.
+  final _stringContentCache = new LruCache<AssetId, String>(
+      1024 * 1024, 1024 * 1024 * 512, (value) => value.length);
+
+  /// Pending `readAsString` operations.
+  final _pendingStringContentCache = <AssetId, Map<Encoding, Future<String>>>{};
```
Comment (natebosch) [inline build_runner_core/lib/src/asset/cache.dart]: why do we still need the nested map here?

Comment (jakemac53) [reply build_runner_core/lib/src/asset/cache.dart]: we don't! good catch

Comment (jakemac53) [reply build_runner_core/lib/src/asset/cache.dart]: fixed

**Pull request merged**